src/base_alloc/base_alloc.c

```diff
@@ -0,0 +1,129 @@
+
+typedef struct ba_mem_chunk_t ba_mem_chunk_t;
+
+// memory chunk of size 'aligned_size'
+struct ba_mem_chunk_t {
+    ba_mem_chunk_t *next;
+    char user_data[];
+};
+
+struct umf_ba_pool_meta_t {
+    size_t aligned_size;       // size of all memory chunks in this pool
+    os_mutex_t *free_lock;     // lock of free_list
+    ba_mem_chunk_t *free_list; // list of free chunks
+};
+
+struct umf_ba_pool_t {
+    struct umf_ba_pool_meta_t metadata;
+    char data[];
+};
+
+// ba_divide_memory_into_chunks - divide given memory into chunks of aligned_size and add them to the free_list
+static void ba_divide_memory_into_chunks(umf_ba_pool_t *pool, void *ptr,
+                                         size_t size) {
+    assert(pool->metadata.free_list == NULL);
+    assert(size > pool->metadata.aligned_size);
+
+    char *pool_memory = ptr;
+    size_t size_left = size;
+
+    ba_mem_chunk_t *current_chunk = (ba_mem_chunk_t *)pool_memory;
+    ba_mem_chunk_t *prev_chunk = current_chunk;
+
+    while (size_left >= pool->metadata.aligned_size) {
+        current_chunk = (ba_mem_chunk_t *)pool_memory;
+        prev_chunk->next = current_chunk;
+
+        pool_memory += pool->metadata.aligned_size;
+        size_left -= pool->metadata.aligned_size;
+        prev_chunk = current_chunk;
+    }
+
+    current_chunk->next = NULL;
+    pool->metadata.free_list = ptr; // address of the first chunk
+}
+
+umf_ba_pool_t *umf_ba_create(size_t size) {
+    umf_ba_pool_t *pool = (umf_ba_pool_t *)ba_os_alloc(BA_POOL_SIZE);
+    if (!pool) {
+        return NULL;
+    }
+
+    char *pool_memory = (char *)&pool->data;
+    size_t size_left = BA_POOL_SIZE - sizeof(struct umf_ba_pool_meta_t);
+
+    // align size of a chunk to MEMORY_ALIGNMENT bytes
+    pool->metadata.aligned_size =
+        align_size(sizeof(ba_mem_chunk_t) + size, MEMORY_ALIGNMENT);
+
+    // allocate and init free_lock
+    pool->metadata.free_lock = util_mutex_init(pool_memory);
+    if (!pool->metadata.free_lock) {
+        return NULL;
+    }
+
+    pool_memory += util_mutex_get_size(); // free_lock is here
+    size_left -= util_mutex_get_size();   // for free_lock
+
+    util_mutex_lock(pool->metadata.free_lock);
+    pool->metadata.free_list = NULL;
+    ba_divide_memory_into_chunks(pool, pool_memory, size_left);
+    util_mutex_unlock(pool->metadata.free_lock);
+
+    return pool;
+}
+
+void *umf_ba_malloc(umf_ba_pool_t *pool) {
+    util_mutex_lock(pool->metadata.free_lock);
+    if (pool->metadata.free_list == NULL) {
+        umf_ba_pool_t *new_pool = (umf_ba_pool_t *)ba_os_alloc(BA_POOL_SIZE);
+        if (!new_pool) {
+            util_mutex_unlock(pool->metadata.free_lock);
+            return NULL;
+        }
+        ba_divide_memory_into_chunks(pool, new_pool, BA_POOL_SIZE);
+    }
+
+    ba_mem_chunk_t *chunk = pool->metadata.free_list;
+    pool->metadata.free_list = pool->metadata.free_list->next;
+    util_mutex_unlock(pool->metadata.free_lock);
+
+    return &chunk->user_data;
```

Comment:
Do we actually need to waste 8 bytes per each allocation? `next` pointer is only used for blocks that are free so can't we just return `(void*) &chunk->next;` here to the user? Or would that be against strict aliasing policy?

Reply:
> Do we actually need to waste 8 bytes per each allocation?

What do you mean is wasted here? A `next` pointer? I'm afraid we can't get rid of it.

Reply:
hm, I think Igor is right. We don't need to get rid of it - it's still needed when the block is free, but can be overwritten with user data when it's given out as allocation. So when you remove  `&chunk->user_data` (and the reverse in free) things should just keep working?

Reply:
Yeah, that was my thinking. In other words, the `ba_mem_chunk_t` could be an union instead of a struct. (but C forbids having a flexible array member in union so we would probably just need to cast the allocation to `ba_mem_chunk_t*` in free).

Reply:
OK, I got it now.
Done.